### PR TITLE
feat(components): Confirmation modal

### PR DESCRIPTION
**Pull request by seanhealy**
### Added

- ConfirmationModal

### Changed

- 

### Deprecated

- 

### Removed

- 

### Fixed

- 

### Security

- 

## Testing

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -147,7 +147,11 @@ module.exports = {
   // ],
 
   // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
-  testPathIgnorePatterns: ["/node_modules/", "/packages/generators/templates/"],
+  testPathIgnorePatterns: [
+    "/node_modules/",
+    "/packages/generators/templates/",
+    ".docz",
```
Comment (seanhealy) [inline jest.config.js:153]: This should be ignored.

File: packages/components/src/ConfirmationModal/ConfirmationModal.css [hunk 2]
```diff
@@ -0,0 +1,10 @@
+.confirmationModal {
```
Comment (seanhealy) [inline packages/components/src/ConfirmationModal/ConfirmationModal.css:1]: This should be removed.

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+- It's good to identify the Object that is being acted upon. For example: "Are
+  you sure you want to delete the user: Bub"
+
+## Props
+
+<Props of={ConfirmationModal} />
+
+## Controlled Confirmation Modal
+
+In some instances, such as a list, you may want to place a single Confirmation
```
Comment (JeffMarvin) [inline packages/components/src/ConfirmationModal/ConfirmationModal.mdx:65]: I think we need a little more instruction here about what the code is below. I.e., how this solution can be used as a template for other solutions.
Is there a way we can make this a component or a function of its own?

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      <>
+        <Button label="Open" onClick={() => setOpen(true)} />
+        <ConfirmationModal
+          title="Should we?"
+          text="Do something…"
+          confirmLabel="Do it"
+          open={open}
+          onConfirm={() => alert("✅")}
+          onCancel={() => alert("🙅‍♂️")}
+          onRequestClose={() => setOpen(false)}
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:41]: Documented in props.

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    );
+  }}
+</Playground>
+
+## Design & Usage Guidelines
+
+ConfirmationModals should be used to confirm or cancel an action the user is
+preforming.
+
+- Ensure that you are labeling the buttons with appropriate verbs. For example:
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:53]: They do.

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      {
+        id: 1,
+        name: "Bob"
+      },
+      {
+        id: 2,
+        name: "Donald"
+      }
+    ];
+    return (
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:81]: It is looping over an array and creating a button for each item in the array. I don't know that it needs a comment… I think it's pretty standard for react template looping?

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+) {
+  switch (action.type) {
+    case "display":
+      return {
+        ...state,
+        title: action.title,
+        text: action.text,
+        open: true,
+        confirmLabel: action.confirmLabel,
+        cancelLabel: action.cancelLabel = "Cancel",
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:53]: Ah, this should be `||`. Fixing.

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+  const [state, dispatch] = useReducer(confirmationModalReducer, {
+    title,
+    text,
+    open: false,
+    confirmLabel,
+    cancelLabel,
+    onConfirm,
+    onCancel,
+  });
+  useImperativeHandle(ref, () => ({
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:133]: `useImperativeHandle` is a hook build into react core as such I'm not sure that it needs a comment but, happy to chat?

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+      },
+    });
+  }, [title, text, confirmLabel, cancelLabel, onConfirm, onCancel]);
+
+  return (
+    <Modal
+      title={state.title}
+      open={open || state.open}
+      size="small"
+      dismissible={false}
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:174]: How so?

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+export interface ConfirmationModalRef {
+  show(props: Omit<DisplayAction, "type">): void;
+}
+
+interface BaseConfirmationModalProps {
+  readonly title?: string;
+  readonly text?: string;
+  readonly open?: boolean;
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:87]: I think `open` in a situation like this is a bit more conventional. In the usage side you are setting open: `<ConfirmationModal open={foo} />`

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx (hunk 3 above)
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:65]: Added some more description here.

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+import { useRef, useState } from "react";
+import { Playground, Props } from "docz";
+import { ComponentStatus } from "@jobber/docx";
+import { Button } from "@jobber/components/Button";
+import { ConfirmationModal } from ".";
+
+# Confirmation Modal
+
+<ComponentStatus stage="pre" responsive="no" accessible="no" />
```
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:15]: Updated `responsive` and `accessible` to mirror Modal as it's what's powering this. Left stage as is for the moment.

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (JeffMarvin) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:98]: Ok, fair.

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:41]: Resolved in convo

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:53]: Resolved in convo

File: packages/components/src/ConfirmationModal/ConfirmationModal.mdx (hunk 6 above)
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.mdx:88]: Resolved in convo

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:164]: Comment added.

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:205]: Resolved in convo

File: packages/components/src/ConfirmationModal/ConfirmationModal.tsx [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+interface ComplexConfirmationModalProps extends BaseConfirmationModalProps {
+  readonly ref: Ref<ConfirmationModalRef>;
+  readonly open?: undefined;
+}
+
+type ConfirmationModalProps =
+  | SimpleConfirmationModalProps
+  | ComplexConfirmationModalProps;
+
+export const ConfirmationModal = forwardRef(ConfirmationModalInternal);
```
Comment (eddysims) [inline packages/components/src/ConfirmationModal/ConfirmationModal.tsx:141]: for OCD reasons, can we move this to the bottom of the file?

Comment (seanhealy) [reply packages/components/src/ConfirmationModal/ConfirmationModal.tsx:141]: Talked through an even better way!

**Comment by eddysims** [comment]
LGTM! 👍

**Pull request merged**